File: framework/dotnet/BetterPortal/Streaming.cs
using AnyVali;
using System.Runtime.CompilerServices;
using System.Text;
using System.Net.ServerSentEvents;

namespace BetterPortal;

public sealed class StreamException(string message) : Exception(message)
{
    public int Status => 500;
}

public sealed record StreamValue<TItem, TSummary>
{
    internal bool IsSummary { get; }
    internal object? Data { get; }
    private StreamValue(bool summary, object? data) { IsSummary = summary; Data = data; }
    public static StreamValue<TItem, TSummary> Item(TItem value) => new(false, value);
    public static StreamValue<TItem, TSummary> Summary(TSummary value) => new(true, value);
}

/// <summary>Request-local producer; each MoveNext validates one value before it reaches a consumer.</summary>
public sealed class StreamHandler<TItem, TSummary, TContext>
{
    public Schema ItemSchema { get; }
    public Schema? SummarySchema { get; }
    public Schema ResponseSchema { get; }
    private readonly Func<TContext, CancellationToken, IAsyncEnumerable<StreamValue<TItem, TSummary>>> run;
    private readonly int maxFrameBytes;

    public StreamHandler(Schema itemSchema, Func<TContext, CancellationToken, IAsyncEnumerable<StreamValue<TItem, TSummary>>> run,
        Schema? summarySchema = null, int maxFrameBytes = 1024 * 1024)
    {
        if (maxFrameBytes < 1024) throw new ArgumentException("Frame limit must be at least 1024 bytes");
        ItemSchema = itemSchema; SummarySchema = summarySchema; this.run = run; this.maxFrameBytes = maxFrameBytes;
        var items = (Dictionary<string, object?>)Json.Read(Json.Write(V.Export(itemSchema)))!;
        items["root"] = new Dictionary<string, object?> { ["kind"] = "array", ["items"] = items["root"], ["default"] = new List<object?>() };
        var properties = new Dictionary<string, Dictionary<string, object?>> { ["items"] = items };
        if (summarySchema is not null)
        {
            var summary = (Dictionary<string, object?>)Json.Read(Json.Write(V.Export(summarySchema)))!;
            summary["root"] = new Dictionary<string, object?> { ["kind"] = "optional", ["inner"] = summary["root"] };
            properties["summary"] = summary;
        }
        ResponseSchema = Contracts.Import(Contracts.ObjectDocument(properties));
    }

    public async IAsyncEnumerable<Dictionary<string, object?>> Frames(TContext context, [EnumeratorCancellation] CancellationToken cancellation = default)
    {
        cancellation.ThrowIfCancellationRequested();
        var producer = run(context, cancellation).GetAsyncEnumerator(cancellation);
        Dictionary<string, object?>? failure = null;
        var count = 0;
        var summarized = false;
        Task<bool>? pending = null;
        try
        {
            while (true)
            {
                Dictionary<string, object?>? frame = null;
                try
                {
                    cancellation.ThrowIfCancellationRequested();
                    pending = producer.MoveNextAsync().AsTask();
                    if (!await pending.WaitAsync(cancellation)) break;
                    cancellation.ThrowIfCancellationRequested();
                    if (summarized) throw new StreamException("Stream produced a value after its summary");
                    var value = producer.Current;
                    var schema = value.IsSummary ? SummarySchema ?? throw new StreamException("Stream has no summary schema") : ItemSchema;
                    frame = new() { ["kind"] = value.IsSummary ? "summary" : "item", ["data"] = Contracts.Parse(schema, value.Data) };
                    if (Encoding.UTF8.GetByteCount(Json.Write(frame)) > maxFrameBytes) throw new StreamException("Stream frame exceeds its byte limit");
                    if (value.IsSummary) summarized = true; else count++;
                }
                catch (OperationCanceledException) when (cancellation.IsCancellationRequested) { throw; }
                catch (ValidationError) { failure = Error("item_validation_failed", "Stream payload validation failed"); }
                catch (Exception) { failure = Error("stream_failed", "Stream failed"); }
                if (failure is not null) break;
                yield return frame!;
            }
        }
        finally
        {
            try
            {
                // An async iterator cannot dispose concurrently with MoveNext.
                // Cancellation releases the caller; cooperative producer I/O releases its resources.
                if (pending is { IsCompleted: false }) _ = CloseAfter(pending, producer);
                else await producer.DisposeAsync();
            }
            catch (OperationCanceledException) when (cancellation.IsCancellationRequested) { throw; }
            catch (Exception) { failure = Error("stream_failed", "Stream failed"); }
        }
        cancellation.ThrowIfCancellationRequested();
        yield return failure ?? (Dictionary<string, object?>)Contracts.Parse("StreamEndFrameSchema", new Dictionary<string, object?> { ["kind"] = "end", ["count"] = count })!;
    }

    private static async Task CloseAfter(Task<bool> pending, IAsyncEnumerator<StreamValue<TItem, TSummary>> producer)
    {
        try { await pending; } catch (Exception) { }
        try { await producer.DisposeAsync(); } catch (Exception) { }
    }

    private static Dictionary<string, object?> Error(string code, string message) =>
        (Dictionary<string, object?>)Contracts.Parse("StreamErrorFrameSchema", new Dictionary<string, object?> { ["kind"] = "error", ["error"] = code, ["message"] = message })!;

    public async Task<Dictionary<string, object?>> Buffered(TContext context, int maxItems = 10000, int maxBytes = 8 * 1024 * 1024, CancellationToken cancellation = default)
    {
        if (maxItems < 0 || maxBytes < 2) throw new ArgumentException("Invalid buffer limits");
        var items = new List<object?>();
        var result = new Dictionary<string, object?> { ["items"] = items };
        long size = Encoding.UTF8.GetByteCount(Json.Write(result));
        await foreach (var frame in Frames(context, cancellation))
        {
            if (frame["kind"] is "error") throw new StreamException((string)frame["message"]!);
            if (frame["kind"] is "item" or "summary")
            {
                size += Encoding.UTF8.GetByteCount(Json.Write(frame["data"]));
                if (frame["kind"] is "item") { size += items.Count > 0 ? 1 : 0; items.Add(frame["data"]); }
                else { size += 11; result["summary"] = frame["data"]; }
                if (items.Count > maxItems || size > maxBytes) throw new StreamException("Stream buffer exceeds its limit");
            }
        }
        if (size > maxBytes) throw new StreamException("Stream buffer exceeds its limit");
        return result;
    }

    public async IAsyncEnumerable<byte[]> Ndjson(TContext context, [EnumeratorCancellation] CancellationToken cancellation = default)
    {
        await foreach (var frame in Frames(context, cancellation)) yield return Encoding.UTF8.GetBytes(Json.Write(frame) + "\n");
    }

    public async IAsyncEnumerable<SseItem<string>> Sse(TContext context, [EnumeratorCancellation] CancellationToken cancellation = default)
    {
        await foreach (var frame in Frames(context, cancellation)) yield return new(Json.Write(frame), (string)frame["kind"]!);
    }
}

File: framework/python/betterportal/streaming.py
"""Validated finite streams with pull backpressure and bounded buffered output."""
from __future__ import annotations

from contextlib import aclosing
from dataclasses import dataclass
import json
from typing import Any, AsyncGenerator, AsyncIterator, Callable, Generic, TypeVar

import anyvali as av
from .contracts import export, object_document, parse
from .sse import encode_event

Item = TypeVar("Item")
SummaryValue = TypeVar("SummaryValue")
Context = TypeVar("Context")


@dataclass(frozen=True)
class Summary(Generic[SummaryValue]):
    value: SummaryValue


class StreamError(ValueError):
    status = 500


def _json(value: Any) -> bytes:
    return json.dumps(value, ensure_ascii=False, allow_nan=False, separators=(",", ":")).encode("utf-8")


class StreamHandler(Generic[Item, SummaryValue, Context]):
    def __init__(self, item_schema: av.BaseSchema[Item],
                 run: Callable[[Context], AsyncIterator[Item | Summary[SummaryValue]]],
                 summary_schema: av.BaseSchema[SummaryValue] | None = None, *, max_frame_bytes: int = 1024 * 1024):
        if max_frame_bytes < 1024:
            raise ValueError("Frame limit must be at least 1024 bytes")
        self.item_schema, self.summary_schema, self.run = item_schema, summary_schema, run
        self.max_frame_bytes = max_frame_bytes
        items = export(item_schema)
        items["root"] = {"kind": "array", "items": items["root"], "default": []}
        properties = {"items": items}
        if summary_schema is not None:
            summary = export(summary_schema)
            summary["root"] = {"kind": "optional", "inner": summary["root"]}
            properties["summary"] = summary
        self.response_schema = av.import_schema(object_document(properties))

    async def frames(self, context: Context) -> AsyncGenerator[dict[str, Any], None]:
        count, summarized = 0, False
        producer = self.run(context).__aiter__()
        failure = None
        try:
            try:
                async for value in producer:
                    if summarized:
                        raise StreamError("Stream produced a value after its summary")
                    if isinstance(value, Summary):
                        if self.summary_schema is None:
                            raise StreamError("Stream has no summary schema")
                        frame = {"kind": "summary", "data": self.summary_schema.parse(value.value)}
                        summarized = True
                    else:
                        frame = {"kind": "item", "data": self.item_schema.parse(value)}
                        count += 1
                    if len(_json(frame)) > self.max_frame_bytes:
                        raise StreamError("Stream frame exceeds its byte limit")
                    yield frame
            except av.ValidationError:
                failure = {"kind": "error", "error": "item_validation_failed", "message": "Stream payload validation failed"}
            except Exception:
                failure = {"kind": "error", "error": "stream_failed", "message": "Stream failed"}
        finally:
            close = getattr(producer, "aclose", None)
            if close is not None:
                try:
                    await close()
                except Exception:
                    failure = {"kind": "error", "error": "stream_failed", "message": "Stream failed"}
        # CancelledError and GeneratorExit propagate: neither produces a terminal frame.
        yield parse("StreamErrorFrameSchema", failure) if failure else parse("StreamEndFrameSchema", {"kind": "end", "count": count})

    async def buffered(self, context: Context, *, max_items: int = 10000, max_bytes: int = 8 * 1024 * 1024) -> dict[str, Any]:
        if max_items < 0 or max_bytes < 2:
            raise ValueError("Invalid buffer limits")
        result: dict[str, Any] = {"items": []}
        size = len(_json(result))
        async with aclosing(self.frames(context)) as frames:
            async for frame in frames:
                if frame["kind"] == "error":
                    raise StreamError(frame["message"])
                if frame["kind"] in ("item", "summary"):
                    if frame["kind"] == "item":
                        size += len(_json(frame["data"])) + (1 if result["items"] else 0)
                        result["items"].append(frame["data"])
                    else:
                        size += len(b',"summary":') + len(_json(frame["data"]))
                        result["summary"] = frame["data"]
                    if len(result["items"]) > max_items or size > max_bytes:
                        raise StreamError("Stream buffer exceeds its limit")
        if size > max_bytes:
            raise StreamError("Stream buffer exceeds its limit")
        return result

    async def ndjson(self, context: Context) -> AsyncGenerator[bytes, None]:
        async with aclosing(self.frames(context)) as frames:
            async for frame in frames:
                yield _json(frame) + b"\n"

    async def sse(self, context: Context) -> AsyncGenerator[bytes, None]:
        async with aclosing(self.frames(context)) as frames:
            async for frame in frames:
                yield encode_event(_json(frame).decode("utf-8"), event=frame["kind"], max_data_bytes=self.max_frame_bytes)
